lib/lint.ml

```diff
@@ -383,6 +453,9 @@ module Check = struct
   let of_dir ~host_os ~master ~job ~packages cwd =
     let master = Current_git.Commit.hash master in
     exec ~cwd ~job [|"git"; "merge"; "-q"; "--"; master|] >>/= fun () ->
+    get_packages ~cwd >>= fun existing_packages ->
+    prev_versions_exist ~cwd packages existing_packages >>= fun prev_versions_exist ->
+    let is_new_package = not prev_versions_exist in
```

Comment:
Iiuc, this will be true if any single package in a PR is new, but couldn't that end up meaning we get several unhelpful checks linting checks if, for instance, one new brand new package is released among several pre-existing packages? E.g., say a new package is factored out of `lwt`, called `lwt-new`, and released along with updates to the `lwt` package. Since the latter didn't exist before, and we are here checking all `packages`, `is_new_package` will be `true`. But then, iiuc, when we linting the `lwt` package, we will end up checking it for name collisions. If my reading of the logic is correct, then it is probably better to move this test into the fold.

Iiuc, a function along these lines should be able to give us a constant time check for whether a single package is being published for the first time:

```ocaml
  let is_newly_published_package ~cwd ~job package master =
    let package_name = OpamPackage.Name.to_string package.OpamPackage.name in
    (* [git cat-file -e branch:path] will exit with zero status if the object at
       [path] exists on [branch] *)
    exec ~cwd ~job [|"git"; "cat-file "; "-e"; master^":"^package_name|]
    >|= function
    | Error _ -> true (* The package directory does not exist on master *)
    | Ok _ -> false (* The package directory does exist on master *)
```

And it could be used as the check for a newly added package on line 475 with something like

```ocaml
          is_newly_published_package ~cwd ~job pkg master >|=
          function
          | false -> errors
          | true ->
              let errors = check_name_collisions ~errors ~pkg existing_packages in
              Prefix.check_name_restricted_prefix ~errors ~pkg
```

In case you like this approach, you can use or reference the patch here: https://github.com/ocurrent/opam-repo-ci/commit/5b67c0a12540ac85932b50906b4e45c0401ce489